server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SortOperator.java

```diff
@@ -67,75 +108,253 @@ public ListenableFuture<?> isBlocked() {
+      }
+    }
+  }
+
+  private void spill() throws IoTDBException {
+    try {
+      // if current memory cannot put this tsBlock, an exception will be thrown in spillSortedData()
+      // because there should be at least tsBlockBuilderStatus.DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES for
+      // one branch.
+      sortBufferManager.allocateOneSortBranch();
+      diskSpiller.spillSortedData(cachedData);
+    } catch (IOException e) {
+      throw new IoTDBException(e.getMessage(), TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
+    }
+  }
+
+  private TsBlock buildTsBlockInMemory() {
+    tsBlockBuilder.reset();
     TimeColumnBuilder timeColumnBuilder = tsBlockBuilder.getTimeColumnBuilder();
     ColumnBuilder[] valueColumnBuilders = tsBlockBuilder.getValueColumnBuilders();
-    cachedData.forEach(
-        mergeSortKey -> {
-          TsBlock tsBlock = mergeSortKey.tsBlock;
-          int row = mergeSortKey.rowIndex;
-          timeColumnBuilder.writeLong(tsBlock.getTimeByIndex(row));
-          for (int i = 0; i < valueColumnBuilders.length; i++) {
-            if (tsBlock.getColumn(i).isNull(row)) {
-              valueColumnBuilders[i].appendNull();
-              continue;
-            }
-            valueColumnBuilders[i].write(tsBlock.getColumn(i), row);
-          }
-          tsBlockBuilder.declarePosition();
-        });
+    for (int i = curRow; i < cachedData.size(); i++) {
+      MergeSortKey mergeSortKey = cachedData.get(i);
+      TsBlock tsBlock = mergeSortKey.tsBlock;
+      timeColumnBuilder.writeLong(tsBlock.getTimeByIndex(mergeSortKey.rowIndex));
+      for (int j = 0; j < valueColumnBuilders.length; j++) {
+        if (tsBlock.getColumn(j).isNull(mergeSortKey.rowIndex)) {
+          valueColumnBuilders[j].appendNull();
+          continue;
+        }
+        valueColumnBuilders[j].write(tsBlock.getColumn(j), mergeSortKey.rowIndex);
+      }
+      tsBlockBuilder.declarePosition();
+      curRow++;
+      if (tsBlockBuilder.isFull()) {
+        break;
+      }
+    }
+
+    return tsBlockBuilder.build();
+  }
+
+  private TsBlock mergeSort() throws IoTDBException {
+
+    if (!diskSpiller.allProcessingTaskFinished()) return null;
+
+    if (mergeSortHeap == null) {
+      mergeSortHeap = new MergeSortHeap(sortReaders.size(), comparator);
+    }
+
+    long startTime = System.nanoTime();
+    long maxRuntime = operatorContext.getMaxRunTime().roundTo(TimeUnit.NANOSECONDS);
+
+    // 1. fill the input from each reader
+    for (int i = 0; i < sortReaders.size(); i++) {
+      if (noMoreData[i] || !isEmpty[i]) continue;
+      SortReader sortReader = sortReaders.get(i);
+      if (sortReader.hasNext()) {
+        MergeSortKey mergeSortKey = sortReader.next();
+        mergeSortKey.tsBlockIndex = i;
+        mergeSortHeap.push(mergeSortKey);
+        isEmpty[i] = false;
+      } else {
+        noMoreData[i] = true;
+        sortBufferManager.releaseOneSortBranch();
+      }
+    }
```

Comment:
Actually, this code block is used to init MergeSortHeap, it can be put into

```
if (mergeSortHeap == null) {
}
```

Reply:
And `if (noMoreData[i] || !isEmpty[i]) continue;` can be deleted.

Reply:
done